x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java

```diff
@@ -0,0 +1,280 @@
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.ql.util.SpatialCoordinateTypes;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static org.apache.lucene.document.ShapeField.QueryRelation.DISJOINT;
+import static org.elasticsearch.xpack.esql.expression.EsqlTypeResolutions.isSpatial;
+import static org.elasticsearch.xpack.esql.expression.function.scalar.spatial.SpatialRelatesUtils.asGeometryDocValueReader;
+import static org.elasticsearch.xpack.esql.expression.function.scalar.spatial.SpatialRelatesUtils.asLuceneComponent2D;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.ql.type.DataTypes.isNull;
+
+public abstract class SpatialRelatesFunction extends BinaryScalarFunction
+    implements
+        EvaluatorMapper,
+        SpatialEvaluatorFactory.SpatialSourceSupplier {
+    protected SpatialCrsType crsType;
+    protected final boolean useDocValues;
+
+    protected SpatialRelatesFunction(Source source, Expression left, Expression right, boolean useDocValues) {
+        super(source, left, right);
+        this.useDocValues = useDocValues;
+    }
+
+    public abstract ShapeField.QueryRelation queryRelation();
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.BOOLEAN;
+    }
+
+    @Override
+    public SpatialCrsType crsType() {
+        if (crsType == null) {
+            resolveType();
+        }
+        return crsType;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        // We determine the spatial data type first, then check if the other expression is of the same type or a string.
+        TypeResolution leftResolution = isSpatial(left(), sourceText(), FIRST);
+        TypeResolution rightResolution = isSpatial(right(), sourceText(), SECOND);
+        // Both are spatial, but one could be a literal spatial function
+        if (leftResolution.resolved() && rightResolution.resolved()) {
+            if (left().foldable() && right().foldable() == false || isNull(left().dataType())) {
+                // Left is literal, but right is not, check the left field's type against the right field
+                return resolveType(right(), left(), FIRST);
+            } else {
+                // All other cases check the right against the left
+                return resolveType(left(), right(), SECOND);
+            }
+        }
+        // If any parameter is not spatial, we are unresolved
+        if ((leftResolution.unresolved() || isNull(left().dataType())) && (rightResolution.unresolved() || isNull(right().dataType()))) {
```

Comment:
This is not supported anymore? let's remove it?

Reply:
We need this for the type error generation. When the function is called with the wrong types, it generates an error message explaining what types it supports and what type the user incorrectly used. These lines are mostly about correct error generation. It is made more complex by the need to support nulls, because if we call it with a null and a spatial type, or two nulls, we don't generate any error, but simply return a null result. But if one type is null, we should generate an error based on the other type. The SpatialIntersectsTests do extensive testing of all possible type combinations and verify that all the correct errors are generated.

Reply:
I wonder if we're better off handling the null cases up front. Honestly this comment helps explain what's happening better than the comments in the code. A little table in the javadoc could help. Something like that.

Also - are we sure that if one of the two is `null` we shouldn't just return `null`? Most other ESQL functions will return `null` if either parameter is `null`. We even have a test utility for it - `anyNullIsNull`. IBM seems to do it that way: https://www.ibm.com/docs/en/ias?topic=functions-st-intersects

That's a lot easier to write too.

Reply:
👍 to Nik's reply to keep things simple - if either side is null, null should be returned.
I understand how the code got here however, extracting the null stuff out will make the code larger but more readable.
Note that there's a subtle difference between null (as in value) vs null as datatype - the latter is rarely used in ESQL and we might forgo it all together.
That is fields always have a type expect except when the user specifies `null` which we could handle through implicit type conversion so that the functions don't have to handle it.